src/common/api/transactions.tsx

```diff
@@ -1,44 +1,156 @@
 import { Transaction } from '@models/transaction.interface';
 import { fetchFromSidecar } from '@common/api/fetch';
-import { TransactionType } from '@blockstack/stacks-blockchain-sidecar-types';
+import { MempoolTransaction, TransactionType } from '@blockstack/stacks-blockchain-api-types';
+import {
+  MempoolTransactionListResponse,
+  SmartContractTransaction,
+  TransactionResults,
+} from '@blockstack/stacks-blockchain-sidecar-types';
+import { generateTypesQueryString } from '@common/api/utils';
+import { fetchContract } from '@common/api/contracts';
+import { ContractDeployTxs, TxData } from '@common/types/tx';
+import { makeKey } from '@common/hooks/use-fetch-blocks';
 
-export const fetchTx = (apiServer: string) => async (
-  txid: Transaction['tx_id']
-): Promise<Transaction> => {
-  const resp = await fetchFromSidecar(apiServer)(`/tx/${txid}`);
-
-  const tx = await resp.json();
-  if (!resp.ok) {
-    throw Error(tx.error);
-  }
-
-  return tx;
-};
+export type FetchTxResponse = Transaction | MempoolTransaction | { error: string };
 
 interface FetchTxListOptions {
   apiServer: string;
-  types: TransactionType[];
+  types?: TransactionType[];
   offset?: number;
+  limit?: number;
+  mempool?: boolean;
 }
 
-interface FetchTxReturnValue {
-  results: Transaction[];
-  total: number;
-}
+export type FetchTransactionResponse =
+  | TxData<Transaction | MempoolTransaction>
+  | { error?: string };
+
+export const fetchTx = (apiServer: string) => async (
+  txid: Transaction['tx_id']
+): Promise<FetchTxResponse> => {
+  const resp = await fetchFromSidecar(apiServer)(`/tx/${txid}`);
+  const tx = await resp.json();
+  return tx as Transaction | MempoolTransaction | { error: string };
+};
 
 export const fetchTxList = (options: FetchTxListOptions) => async (): Promise<
-  FetchTxReturnValue
+  TransactionResults | MempoolTransactionListResponse
 > => {
-  const { apiServer, types, offset } = options;
-  const generateTypesQueryString = () => {
-    if (types?.length) {
-      return types
-        .map(type => `${encodeURIComponent('type[]')}=${encodeURIComponent(type)}`)
-        .join('&');
-    }
-  };
+  const { apiServer, types, offset, limit = 200, mempool } = options;
+
   const resp = await fetchFromSidecar(apiServer)(
-    `/tx?${generateTypesQueryString()}&limit=200${offset ? '&offset=' + offset : ''}`
+    `/tx${mempool ? '/mempool' : ''}?limit=${limit}${
+      offset ? `&offset=${offset}` : ''
+    }${generateTypesQueryString(types)}`
   );
   return resp.json();
 };
+
```

Comment:
You can use Typescript function signature overloads here, which makes things much nicer! Then, the return type will _only_ be `Transaction` or `MempoolTransaction`, not either one.

```ts
interface FetchConfirmedTxListOptions extends FetchTxListOptions {
  mempool?: false;
}

interface FetchPendingTxListOptions extends FetchTxListOptions {
  mempool?: true;
}

export function fetchTxList(
  options: FetchConfirmedTxListOptions
): () => Promise<TransactionResults>;
export function fetchTxList(
  options: FetchPendingTxListOptions
): () => Promise<MempoolTransactionListResponse>;
export function fetchTxList(
  options: FetchTxListOptions
): () => Promise<TransactionResults | MempoolTransactionListResponse> {
  return async () => {
    const { apiServer, types, offset, limit = 200, mempool } = options;

    const resp = await fetchFromSidecar(apiServer)(
      `/tx${mempool ? '/mempool' : ''}?limit=${limit}${
        offset ? `&offset=${offset}` : ''
      }${generateTypesQueryString(types)}`
    );
    return resp.json();
  };
}
```